### ISSUE TITLE
Error reading multiple variables from a YAML File....

### ISSUE THREAD

**Issue by dinooz** [leafty/node-yaml-config#1]
Trying to build a reader of YAML for a WebApp builder.
[ https://groups.google.com/d/msg/nodejs/SBLPR7dYMfg/eRT2uCyhUDQJ ]

Start with a simple YAML File...

default:
  server:
    port: 3000
  database:
    host: 'localhost'
    port: 27017
development:
  database:
    db: 'dev_db'
test:
  database:
    db: 'test_db'
production:
  server:
    port: 8000
  database:
    db: 'prod_db'
    user: 'dbuser'
    password: 'pass'
  cache:
    dir: 'static'

Using the following code in a .js file:

var yaml_config = require('node-yaml-config');

var config = yaml_config.load('myconfig.yml');

console.log(config); 

console.log('Port: ' + config.server.port); 
console.log('dB: ' + config.database.db); 
console.log('Port: ' + config.database.port); 
console.log('Host: ' + config.database.host); 

console.log('Development dB: ' + config.development.db); 

Work for the First set of the config. but return error when trying to list the second variables...

c:\nodejs\dinooz\reader.js:12
console.log('Development dB: ' + config.development.db);
                                                   ^
TypeError: Cannot read property 'db' of undefined
    at Object.<anonymous> (c:\nodejs\dinooz\reader.js:12:52)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:902:3

I wonder if there is some special flag to call the nested variables. The idea is to re-use all the variables for the WebApp builder.

THX Dinooz.

**Comment by leafty** [comment]
You don't have access to the first level of the yaml document, as it is used to specify the environment you are describing.

If in `app.js` you have:

``` js
var yaml_config = require('node-yaml-config');

var config = yaml_config.load(__dirname + '/config/config.yml');

console.log(config); 
```

Then, when you run `node app.js` you'll see:

``` js
{ server: { port: 3000 },
  database: { host: 'localhost', port: 27017, db: 'dev_db' } }
```

This is the default configuration with the `development` config merged in, since the default node environment is `development`. Now, this is a basic javascript object, so you can access its properties like any other object, e.g. by doing `config.server.port` or `config.database.db`.

If you want to use your app in production, just write `NODE_ENV=production node app.js`, and then you'll see:

``` js
{ server: { port: 8000 },
  database: 
   { host: 'localhost',
     port: 27017,
     db: 'prod_db',
     user: 'dbuser',
     password: 'pass' },
  cache: { dir: 'static' } }
```

You can also force the environment to load:

``` js
var config = yaml_config.load(__dirname + '/config/config.yml', 'test');
```

In short, you write the configuration under `default`, and when environment specific variables are needed, you add or override them in the `development`, `test` or `production` yaml node.

**Comment by dinooz** [comment]
I think you are missing the point... I'm not interested in to run the config with Multiple Environment variables. I'm thinking to put all the configurations of the web applications, pages, views, look & feel, models, relationships using this file. my problem is how to separate the YAML file with each of those things and parse it properly. the usage os ENV TEST PROD, is just the begining of how things will be listed/documented, but much more data will be added to this file.

Imagine a Cisco Like text File for the configuration....

Version 1.0 
site
  title
  url
  logo
  slogan
  admin_name
  admin_email
!
  lang
    en,sp
!
  timezone
!  
  layout
!  
  theme
!
  header
!
  footer
!
  banner
!
  nav
!
app
  resources
!
  model
!
  view
!
  controller
!
  analytics
!
end

But like to separate... perhaps some settings are for:
environments [ Devel - test - Production ]
Navigation [ Home - Products - Services - AboutUs ]
Models [ .... web app models ]
Views [ ... Views of the pages ]
Controllers [ ...Controllers of the  pages ]
Look & Feel [ ... define the look and feel of the site Templates/Skin/Theme ]

But I like to have separated in multiple methods within the YAML File this way can call independently.

THX Dino.

**Comment by leafty** [comment]
I'm sorry, I don't think I understand what you are trying to do. The elements that do not vary depending on the environment are to be put under default so that you will retrieve them anyway. It may look like this:

``` yaml
default:
    navigation:
        home: <something>
        products: <something>
        services: <something>
        about: <something>
    models: [...]
    [...]

development:
    server:
        port: 3000
    [...]

test: [...]

production: [...]
```

If you want to read and write YAML documents, take a look at the [js- yaml](https://github.com/nodeca/js-yaml) package. It's quite simple to use. And sorry about the delay.
